Agora que você já dominou as habilidades básicas
necessárias para escrever programas organizados, fáceis
de usar, é hora de pensar em deixar seus programas mais
relevantes e utilizáveis. Neste capítulo aprenderemos a
trabalhar com arquivos para que seus programas possam
analisar rapidamente muitos dados.
Veremos como tratar erros a fim de que seus programas não falhem
quando se depararem com situações inesperadas. Conheceremos as
exceções – objetos Python especiais para administrar erros que surgirem
enquanto um programa estiver executando. Também conheceremos o
módulo json, que permite salvar dados de usuário para que não sejam
perdidos quando seu programa parar de executar.
Aprender a trabalhar com arquivos e a salvar dados deixará seus
programas mais fáceis de usar. Os usuários poderão escolher quais dados
devem fornecer e quando. As pessoas podem executar seu programa, fazer
alguma tarefa e então fechá-lo e retomá-lo mais tarde, do ponto em que
pararam. Aprender a tratar exceções ajudará você a lidar com situações em
que os arquivos não existam e com outros problemas que possam fazer
seus programas falharem. Isso deixará seus programas mais robustos
quando dados ruins forem encontrados, sejam eles provenientes de erros
inocentes ou de tentativas maliciosas de fazer seus programas falharem.
Com as habilidades desenvolvidas neste capítulo, você deixará seus
programas mais aplicáveis, utilizáveis e estáveis.

## Lendo dados de um arquivo
Uma quantidade incrível de dados está disponível em arquivos-texto. Os
arquivos-texto podem conter dados meteorológicos, de tráfego,
socioeconômicos, trabalhos literários e outros. Ler dados de um arquivo é
particularmente útil em aplicações de análise de dados, mas também se
aplica a qualquer situação em que você queira analisar ou modificarinformações armazenadas em um arquivo. Por exemplo, podemos escrever
um programa que leia o conteúdo de um arquivo-texto e reescreva o
arquivo com uma formatação que permita a um navegador exibi-lo.
Quando quiser trabalhar com as informações de um arquivo-texto, o
primeiro passo será ler o arquivo em memória. Você pode ler todo o
conteúdo de um arquivo ou pode trabalhar com uma linha de cada vez.

## Lendo um arquivo inteiro

In [2]:
with open('data/pi_digits.txt') as file_object:
    contents = file_object.read()
    print(contents)

3.1415926535
8979323846
2643383279


Muitas atividades acontecem na primeira linha desse programa. Vamos
começar observando a função open(). Para realizar qualquer tarefa com um
arquivo, mesmo que seja apenas exibir o seu conteúdo, você precisará
inicialmente abrir o arquivo para acessá-lo. A função open() precisa de um
argumento: o nome do arquivo que você quer abrir. Python procura esse
arquivo no diretório em que o programa executando no momento estáarmazenado. Nesse exemplo, file_reader.py está executando, portanto
Python procura pi_digits.txt no diretório em que o arquivo file_reader.py
está armazenado. A função open() devolve um objeto que representa o
arquivo. Nesse caso, open('pi_digits.txt') devolve um objeto que
representa pi_digits.txt. Python armazena esse objeto em file_object, com
o qual trabalharemos posteriormente no programa.
A palavra reservada with fecha o arquivo depois que não for mais
necessário acessá-lo. Observe como chamamos open() nesse programa, mas
não chamamos close(). Você poderia abrir e fechar o arquivo chamando
open() e close(), mas se um bug em seu programa impedir que a instrução
close() seja executada, o arquivo não será fechado. Isso pode parecer
trivial, mas arquivos indevidamente fechados podem provocar perda de
dados ou estes podem ser corrompidos. Além disso, se close() for
chamado cedo demais em seu programa, você se verá tentando trabalhar
com um arquivo fechado (um arquivo que não pode ser acessado), o que
resultará em mais erros. Nem sempre é fácil saber exatamente quando
devemos fechar um arquivo, mas com a estrutura mostrada aqui, Python
descobrirá isso para você. Tudo que você precisa fazer é abrir o arquivo e
trabalhar com ele conforme desejado, com a confiança de que Python o
fechará automaticamente no momento certo.
Depois que tivermos um objeto arquivo que represente pi_digits.txt,
usamos o método read() na segunda linha de nosso programa para ler todo
o conteúdo do arquivo e armazená-lo em uma longa string em contents.
Quando exibimos o valor de contents, vemos o arquivo-texto completo:
3.1415926535
8979323846
2643383279
A única diferença entre essa saída e o arquivo original é a linha em branco
extra no final da saída. A linha em branco aparece porque read() devolve
uma string vazia quando alcança o final do arquivo; essa string vazia
aparece como uma linha em branco. Se quiser remover essa linha em
branco extra, rstrip() pode ser usada na instrução print:

In [3]:
with open('data/pi_digits.txt') as file_object:
    contents = file_object.read()
    print(contents.rstrip())

3.1415926535
8979323846
2643383279


### Paths de arquivo
Quando um nome de arquivo simples como pi_digits.txt é passado para a
função open(), Python observa o diretório em que o arquivo executado no
momento (isto é, seu arquivo de programa .py) está armazenado.
Às  vezes,  dependendo  de  como  o  seu  trabalho  estiver  organizado,  o
arquivo  que  você  quer  abrir  não  estará  no  mesmo  diretório  que  o  seu
arquivo de programa. Por exemplo, você pode armazenar seus arquivos de
programa em uma pasta chamada python_work; nessa pasta pode haver
outra pasta chamada text_files para distinguir seus arquivos de programa
dos  arquivos-texto  que  eles  manipulam.  Apesar  de  text_files  estar  em
python_work, simplesmente passar o nome de um arquivo que está em
text_files  para  open()  não  funcionará,  pois  Python  procurará  o  arquivo
apenas  em  python_work;  ele  não  prosseguirá  procurando  em  text_files.
Para fazer Python abrir arquivos de um diretório que não seja aquele em
que seu arquivo de programa está armazenado, é preciso fornecer um path
de arquivo, que diz a Python para procurar em um local específico de seu
sistema.
Como  text_files  está  em  python_work,  você  pode  usar  um  path  de
arquivo relativo para abrir um arquivo em text_files. Um path de arquivo
relativo  diz  a  Python  para  procurar  um  local  especificado,  relativo  ao
diretório em que o arquivo de programa em execução no momento está
armazenado. No Linux e no OS X, você escreveria o seguinte:
with open('text_files/nome_do_arquivo.txt') as file_object:
Essa linha diz a Python para procurar o arquivo .txt desejado na pasta
text_files e supõe que essa pasta está localizada em python_work (e está).
Em sistemas Windows, use uma barra invertida (\) no lugar da barra para a
frente (/) no path do arquivo:
with open('text_files\nome_do_arquivo.txt') as file_object:
Você  também  pode  dizer  a  Python  exatamente  em  que  local  está  o
arquivo em seu computador, não importando o lugar em que o programa
em  execução  no  momento  esteja  armazenado.  Isso  é  chamado  de  path
absoluto do arquivo. Utilize um path absoluto se um path relativo não
funcionar.  Por  exemplo,  se  você  colocou  text_files  em  alguma  pasta
diferente  de  python_work  –  por  exemplo,  em  uma  pasta  chamada
other_files  –  então  simplesmente  passar  o  path
'text_files/nome_do_arquivo.txt'  para  open()  não  funcionará  porque
Python  procurará  esse  local  somente  em  python_work.  Você  precisará
fornecer um path completo para deixar claro em que lugar você quer que
Python procure.
Paths absolutos geralmente são mais longos que paths relativos, portanto
é conveniente armazená-los em uma variável e então passar essa variável
para open(). No Linux e no OS X, paths absolutos têm o seguinte aspecto:
file_path =
'/home/ehmatthes/other_files/text_files/nome_do_arquivo.txt'
with open(file_path) as file_object:
No Windows, eles se parecem com:
file_path =
'C:\Users\ehmatthes\other_files\text_files\nome_do_arquivo.txt'
with open(file_path) as file_object:
Ao  usar  paths  absolutos,  podemos  ler  arquivos  de  qualquer  lugar  do
sistema. Por enquanto, é mais fácil armazenar arquivos no mesmo diretório
em que estão seus arquivos de programa ou em uma pasta como text_files
no diretório em que estiverem seus arquivos de programa.
NOTA Sistemas Windows às vezes interpretam corretamente as barras para a frente
nos paths de arquivo. Se você usa Windows e não está obtendo os resultados
esperados, tente usar barras invertidas.

### Lendo dados linha a linha
Quando estiver lendo um arquivo, com frequência você vai querer analisar
cada  linha  do  arquivo.  Talvez  você  esteja  procurando  determinada
informação no arquivo ou queira modificar o texto do arquivo de alguma
maneira. Por exemplo, você pode ler um arquivo de dados meteorológicos
e  trabalhar  com  qualquer  linha  que  inclua  a  palavra  ensolarado  na
descrição da previsão do tempo para esse dia. Em uma notícia, talvez você
queira procurar todas as linhas com a tag <headline> e reescrever essa linha
com um tipo específico de formatação.
Podemos usar um laço for no objeto arquivo para analisar cada uma de
suas linhas, uma de cada vez:

In [ ]:
filename = 'data/pi_digits.txt'
with open(filename) as file_object:
    for line in file_object:
        print(line.rstrip())
        """ The rstrip() method removes any whitespace characters from the end of the string, including the newline character. This way, when we print each line, it won't have an extra blank line after it. """

3.1415926535
8979323846
2643383279


### Criando uma lista de linhas de um arquivo
Quando  usamos  with,  o  objeto  arquivo  devolvido  por  open()  estará
disponível somente no bloco with que o contém. Se quiser preservar o
acesso  ao  conteúdo  de  um  arquivo  fora  do  bloco  with,  você  pode
armazenar  as  linhas  do  arquivo  em  uma  lista  dentro  do  bloco  e  então
trabalhar com essa lista. Pode processar partes do arquivo imediatamente e
postergar parte do processamento de modo que seja feito mais tarde no
programa.
O exemplo a seguir armazena as linhas de pi_digits.txt em uma lista no
bloco with e, em seguida, exibe as linhas fora desse bloco:

In [2]:
filename = 'data/pi_digits.txt'

with open(filename) as file_object:
    content = file_object.readlines()

for line in content:
    print(line.rstrip())

3.1415926535
8979323846
2643383279


### Trabalhando com o conteúdo de um arquivo
Depois de ler um arquivo em memória, você poderá fazer o que quiser com
esses dados; desse modo, vamos explorar rapidamente os dígitos de pi. Em
primeiro  lugar,  tentaremos  criar  uma  única  string  contendo  todos  os
dígitos do arquivo, sem espaços em branco:

In [9]:
filename = 'data/pi_digits.txt'
with open(filename) as file_object:
    lines = file_object.readlines()

for line in lines:
    print(line.rstrip(), end='')
    
""" ou... """
print()
pistring = ''

for line in lines:
    pistring += line.strip()

print(pistring)
print(len(pistring))

3.141592653589793238462643383279
3.141592653589793238462643383279
32


### Arquivos grandes: um milhão de dígitos
Até  agora,  nosso  enfoque  foi  analisar  um  arquivo-texto  que  continha
apenas três linhas, mas o código desses exemplos também funcionará bem
em arquivos muito maiores. Se começarmos com um arquivo-texto que
contenha pi com um milhão de casas decimais, e não trinta, uma única
string  contendo  todos  esses  dígitos  poderá  ser  criada.  Não  precisamos
alterar  nada  em  nosso  programa,  exceto  para  lhe  passar  um  arquivo
diferente. Além disso, exibiremos as cinquenta primeiras casas decimais
para que não seja necessário assistirmos a um milhão de dígitos rolando
pelo terminal:

In [11]:
filename = 'data/pinums.txt'
with open(filename) as file_object:
    pistring = ''
    content = file_object.readlines()

for line in content:
    pistring += line.strip()
    
print(pistring[:52])
print(len(pistring))

3,14159265358979323846264338327950288419716939937510
1002


In [ ]:
filename = 'data/pinums.txt'
with open(filename) as file_object:
    pistring = ''
    content = file_object.readlines()

for line in content:
    pistring += line.strip()
    
print(pistring[:52])
print(len(pistring))

birthday = input("Enter your birthday, in the form mmddyy: ")
if birthday in pistring:
    print("Your birthday appears in the first thousand digits of pi!")
else:
    print("Your birthday does not appear in the first thousand digits of pi.")

3,14159265358979323846264338327950288419716939937510
1002
Your birthday does not appear in the first million digits of pi.


## FAÇA VOCÊ MESMO
10.1 – Aprendendo Python: Abra um arquivo em branco em seu editor de texto e
escreva  algumas  linhas  que  sintetizem  o  que  você  aprendeu  sobre  Python  até
agora.  Comece  cada  linha  com  a  expressão  Em  Python  podemos....  Salve  o
arquivo como learning_python.txt no mesmo diretório em que estão seus exercícios
deste  capítulo.  Escreva  um  programa  que  leia  o  arquivo  e  mostre  o  que  você
escreveu, três vezes. Exiba o conteúdo uma vez lendo o arquivo todo, uma vez
percorrendo o objeto arquivo com um laço e outra armazenando as linhas em uma
lista e então trabalhando com ela fora do bloco with.

10.2  –  Aprendendo  C:  Você  pode  usar  o  método  replace()  para  substituir
qualquer palavra por uma palavra diferente em uma string. Eis um exemplo rápido
que mostra como substituir a palavra 'dog' por 'cat' em uma frase:
>>> message = "I really like dogs."
>>> message.replace('dog', 'cat')
'I really like cats.'
Leia  cada  linha  do  arquivo  learning_python.txt  que  você  acabou  de  criar  e
substitua a palavra Python pelo nome de outra linguagem, por exemplo, C. Mostre
cada linha modificada na tela.


In [6]:
with open('data/learning_python.txt') as file_object:
    lines = file_object.readlines()
    content = file_object.read()

count = 3
while count > 0:
    for line in lines:
        print(line.strip())
    print()
    count -= 1
    
print(f'{content}\n')

for line in lines:
    print(line.strip())
print()


# learning_python.txt

Em Python podemos criar variáveis sem declarar o tipo de dado.
Em Python podemos usar listas para armazenar múltiplos valores.
Em Python podemos criar funções reutilizáveis com a palavra-chave def.
Em Python podemos iterar sobre sequências com loops for e while.
Em Python podemos trabalhar com dicionários para armazenar dados em pares chave-valor.
Em Python podemos usar condicionais if, elif e else para controlar o fluxo do programa.
Em Python podemos manipular strings com métodos integrados como lower(), upper() e split().
Em Python podemos abrir e ler arquivos para processar dados.
Em Python podemos organizar código em módulos e importá-los conforme necessário.
Em Python podemos tratar erros usando blocos try e except.

# learning_python.txt

Em Python podemos criar variáveis sem declarar o tipo de dado.
Em Python podemos usar listas para armazenar múltiplos valores.
Em Python podemos criar funções reutilizáveis com a palavra-chave def.
Em Python podemos iterar

In [8]:
with open('data/learning_python.txt') as file_object:
    content = file_object.read()
    swaped = content.replace('Python', 'C')
    print(swaped)

# learning_python.txt

Em C podemos criar variáveis sem declarar o tipo de dado.
Em C podemos usar listas para armazenar múltiplos valores.
Em C podemos criar funções reutilizáveis com a palavra-chave def.
Em C podemos iterar sobre sequências com loops for e while.
Em C podemos trabalhar com dicionários para armazenar dados em pares chave-valor.
Em C podemos usar condicionais if, elif e else para controlar o fluxo do programa.
Em C podemos manipular strings com métodos integrados como lower(), upper() e split().
Em C podemos abrir e ler arquivos para processar dados.
Em C podemos organizar código em módulos e importá-los conforme necessário.
Em C podemos tratar erros usando blocos try e except.


### Escrevendo dados em um arquivo
Uma  das  maneiras  mais  simples  de  salvar  dados  é  escrevê-los  em  um
arquivo. Quando um texto é escrito em um arquivo, o resultado estará
disponível depois que você fechar o terminal que contém a saída de seu
programa. Podemos analisar a saída depois que um programa acabar de
executar e compartilhar os arquivos de saída com outras pessoas também.
Além disso, podemos escrever programas que leiam o texto de volta para a
memória e trabalhar com esses dados novamente.

### Escrevendo dados em um arquivo vazio
Para escrever um texto em um arquivo, chame open() com um segundo
argumento que diga a Python que você quer escrever dados no arquivo.
Para ver como isso funciona, vamos escrever uma mensagem simples e
armazená-la em um arquivo em vez de exibi-la na tela:

In [2]:
filename = 'data/programming.txt'

with open(filename, 'w') as file_object:
    file_object.write("I love programming.")

A chamada a open() nesse exemplo tem dois argumentos u. O primeiro
argumento ainda é o nome do arquivo que queremos abrir. O segundo
argumento, 'w', diz a Python que queremos abrir o arquivo em modo de
escrita. Podemos abrir um arquivo em modo de leitura ('r'), em modo de
escrita ('w'), em modo de concatenação ('a') ou em um modo que permita
ler e escrever no arquivo ('r+'). Se o argumento de modo for omitido, por
padrão Python abrirá o arquivo em modo somente de leitura.
A função open() cria automaticamente o arquivo no qual você vai escrever
caso ele ainda não exista. No entanto, tome cuidado ao abrir um arquivo
em modo de escrita ('w') porque se o arquivo já existir, Python o apagará
antes de devolver o objeto arquivo.
Em v usamos o método write() no objeto arquivo para escrever uma
string nesse arquivo. Esse programa não tem saída no terminal, mas se
abrir o arquivo programming.txt, você verá uma linha:
programming.txt
I love programming.
Esse  arquivo  se  comporta  como  qualquer  outro  arquivo  de  seu
computador. Você pode abri-lo, escrever um novo texto nele, copiar ou
colar dados e assim por diante.

## NOTA 
Python escreve apenas strings  em um arquivo-texto. Se quiser armazenar
dados numéricos em um arquivo-texto, será necessário converter os dados em
um formato de string antes usando a função str().

### Escrevendo várias linhas
A função write() não acrescenta nenhuma quebra de linha ao texto que
você  escrever.  Portanto,  se  escrever  mais  de  uma  linha  sem  incluir
caracteres  de  quebra  de  linha,  seu  arquivo  poderá  não  ter  a  aparência
desejada:

In [7]:
filename = 'data/programming.txt'
with open(filename, 'r+') as file_object:
    file_object.write('I love programming.')
    file_object.write('I love creating new games.')

with open(filename, 'r') as file_object:
    content = file_object.read()
    print(content)

I love programming.I love creating new games.


A inclusão de quebras de linha em suas instruções write() faz cada string
aparecer em sua própria linha:

In [9]:
filename = 'data/programming.txt'
with open(filename, 'w') as file_object:
    file_object.write("I love programming.\n")
    file_object.write("I love creating new games.\n")

with open(filename) as file_object:
    content = file_object.read()
    print(content)

I love programming.
I love creating new games.



Podemos  também  usar  espaços,  caracteres  de  tabulação  e  linhas  em
branco para formatar a saída, exatamente como viemos fazendo com as
saídas no terminal.

### Concatenando dados em um arquivo
Se quiser acrescentar conteúdos em um arquivo em vez de sobrescrever o
conteúdo existente, você pode abrir o arquivo em modo de concatenação.
Ao abrir um arquivo em modo de concatenação, Python não apagará o
arquivo  antes  de  devolver  o  objeto  arquivo.  Qualquer  linha  que  você
escrever no arquivo será adicionada no final. Se o arquivo ainda não existe,
Python criará um arquivo vazio para você.
Vamos modificar write_message.py acrescentando alguns novos motivos
pelos quais amamos programar no arquivo existente programming.txt:

In [10]:
filename = 'data/programming.txt'
with open(filename, 'a') as file_object:
    file_object.write("I also love finding meaning in large datasets.\n")
    file_object.write("I love creating apps that can run in a browser.\n")
with open(filename) as file_object:
    content = file_object.read()
    print(content)

I love programming.
I love creating new games.
I also love finding meaning in large datasets.
I love creating apps that can run in a browser.



## FAÇA VOCÊ MESMO
10.3  –  Convidado:  Escreva  um  programa  que  pergunte  o  nome  ao  usuário.
Quando ele responder, escreva o nome em um arquivo chamado guest.txt.
10.4  –  Lista  de  convidados: Escreva um laço while que pergunte o nome aos
usuários.  Quando  fornecerem  seus  nomes,  apresente  uma  saudação  na  tela  e
acrescente  uma  linha  que  registre  a  visita  do  usuário  em  um  arquivo  chamado
guest_book.txt. Certifique-se de que cada entrada  esteja em uma nova linha do
arquivo.
10.5  –  Enquete  sobre  programação:  Escreva  um  laço  while  que  pergunte  às
pessoas por que elas gostam de programação. Sempre que alguém fornecer um
motivo, acrescente-o em um arquivo que armazene todas as respostas.

In [11]:
name = input("Escreva seu nome: ")

with open('data/guests.txt', 'a') as file_object:
    file_object.write(f'{name}\n')    

In [12]:
while True:
    name = input("Escreva 'sair' a qualquer momento para encerrar. Escreva seu nome: ")
    if name.lower() == 'sair':
        break
    print(f'Olá, {name}! Seu nome foi adicionado ao livro de visitas.')
    with open('data/guest_book.txt', 'a') as file_object:
        file_object.write(f'{name}\n')

Olá, joão! Seu nome foi adicionado ao livro de visitas.
Olá, carla! Seu nome foi adicionado ao livro de visitas.


In [19]:
while True:
    name = input("Escreva 'sair' a qualquer momento para encerrar. Escreva seu nome: ")
    if name.lower() == 'sair':
        break
    answer = input("Por que você gosta de programar? ")
    if answer.lower() == 'sair':
        break
    with open('data/programming_poll.txt', 'a') as file_object:
        file_object.write(f'{name}: {answer}\n')
with open('data/programming_poll.txt') as file_object:
    content = file_object.read()
    print(content)


joão: por causa dos jogos

